src/compile/selection/assemble.ts

```diff
@@ -203,7 +204,7 @@ export function assembleSelectionScaleDomain(model: Model, domainRaw: SignalRef)
   return {signal: 'null'};
 }
 
-function cleanupEmptyOnArray(signals: Signal[]) {
+function cleanupEmptyOnArray(signals: readonly Signal[]) {
```

Comment:
Another case here, this function is mutating the objects within the array. So labeling the array as `readonly` feels like it's indicating a true rather than shallow immutability. I worry that it gives us a false sense of security/purity when there is more work to be done actually making things pure.

Reply:
Good catch. I'll remove the annotation here.